folders.go

```diff
@@ -0,0 +1,91 @@
+package gapi
+
+import (
+	"bytes"
+	"encoding/json"
+
+	"fmt"
+	"io/ioutil"
+
+	"github.com/pkg/errors"
+)
+
+
+type FolderCreateResponse struct {
+	Id     int `json:"id"`
+	uid    string `json:"uid"`
+	title  string `json:"title"`
+}
+
+type Folder struct {
+	//Model map[string]interface{} `json:"folder"`
+	Id int `json:"id"`
+}
+
+func (c *Client) CreateFolder(model map[string]interface{}) (*FolderCreateResponse, error) {
+	wrapper := map[string]interface{}{
+		"title": model["title"],
+		"uid": 	 model["uid"],
+	}
+	data, err := json.Marshal(wrapper)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to marshall folder JSON")
+	}
+	req, err := c.newRequest("POST", "/api/folders", nil, bytes.NewBuffer(data))
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to perform HTTP request")
+	}
+
+	if resp.StatusCode != 200 {
+		var gmsg GrafanaErrorMessage
+		dec := json.NewDecoder(resp.Body)
+		dec.Decode(&gmsg)
+		return nil, fmt.Errorf("Request to Grafana returned %+v status code with the following message: %+v", resp.StatusCode, gmsg.Message)
+	}
+
+	data, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	result := &FolderCreateResponse{}
+	err = json.Unmarshal(data, &result)
+	return result, err
+}
+
+func (c *Client) Folder(slug string) (*Folder, error) {
+	path := fmt.Sprintf("/api/folders/%s", slug)
+	req, err := c.newRequest("GET", path, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the error code is 404 that means that the folder does not exist. Don't treat this case as an error
```

Comment:
Because if we are actually calling this function, its because we are making a commit with a folder/dashboard.json 
That means that we WANT that folder to be there
404 is the response of the API for a non-existing resource, but as we actually want it to be there, 404 is the response that for us is non an error, its the "folder its not there, you should create it" and let cuarzo know that proceed with the creation process

Reply:
You're tying this too much to cuarzo when in fact this is a generic client for the grafana API.  There are several issues here: 

1. The purpose of this function is to purely GET a folder based on it's UID ( you should change the slug variable name). It's merely requesting an existing resource - this function shouldn't care about what the intentions of the caller are. All it does is return a folder object or an error. The caller is responsible for handling the error as appropriate. In our use case - we have to explicitly handle a 404 error as it means the folder we want doesn't exist and then call the CreateFolder function above to create it. By suppressing the 404 status code we lose the ability to definitively check for the folder presence - since all that's returned is nil, and that could happen in other cases as well.

Reply:
ok, I assume we can change that. 
I dont see that the cuarzo code if we return the err

Reply:
Just added the error handling on latest commit